### PR TITLE
Update MQTT

### PR DESCRIPTION
**Pull request by WillB97**
Fixes the issue with make_safe not running when killed from the webUI when used with https://github.com/srobo/astoria/pull/180.

Decouples the MQTT section of Astoria from sr-robot3, the message structures are still imported from Astoria.
To avoid the issues surrounding appropriate methods to do async event loop clean up in a largely synchronous codebase, this moves to using the paho-mqtt library which uses a background thread to handle the network loop. Cleanup is then handled by an atexit handler to disconnect and close the network loop thread.

To keep it simple there is a single long running client that is created within the robot's init routine and metadata and wait_start are done by temporarily subscribing to the relevant topics. Timeouts are used on the threading Events to prevent indefinite blocking.

### CODE REVIEW

File: .flake8 [hunk 1]
```diff
@@ -13,5 +13,7 @@ ignore =
     D107
     #Don't require imperative mood docstrings.
     D401
+    #Allow binary operators at the beginning of lines
+    W503
```
Comment (trickeydan) [inline .flake8]: It's better to use line continuations than do this, see my in-line suggestions below:

File: sr/robot3/astoria.py [hunk 2]
```diff
@@ -28,126 +31,197 @@ class GetMetadataResult(NamedTuple):
     usb_path: Path
 
 
-class GetMetadataConsumer(StateConsumer):
-    """Astoria consumer to fetch metadata."""
+class GetMetadataConsumer():
```
Comment (trickeydan) [inline sr/robot3/astoria.py]:
```suggestion
class GetMetadataConsumer:
```

Brackets no longer necessary

File: sr/robot3/astoria.py [hunk 3]
```diff
@@ -28,126 +31,197 @@ class GetMetadataResult(NamedTuple):
...
-        logging.getLogger("gmqtt").setLevel(logging.WARNING)
+    def __init__(self, mqtt_client: MQTTClient) -> None:
+        self._mqtt = mqtt_client
 
-    def _init(self) -> None:
-        """Initialise consumer."""
+        # Initialise consumer data structures.
         self._metadata_message: Optional[MetadataManagerMessage] = None
         self._proc_message: Optional[ProcessManagerMessage] = None
-        self._state_lock = asyncio.Lock()
```
Comment (trickeydan) [inline sr/robot3/astoria.py:45]: Any reason we're dropping the lock on the data? It looks like `loop_start` uses a thread pool so the lock is still relevant?

File: sr/robot3/astoria.py [hunk 4]
```diff
@@ -28,126 +31,197 @@ class GetMetadataResult(NamedTuple):
...
+            message = parse_obj_as(MetadataManagerMessage, loads(msg.payload))
+            if message.status == MetadataManagerMessage.Status.RUNNING:
+                LOGGER.debug("Received metadata")
+                self._metadata_message = message
+            else:
+                LOGGER.warn("Cannot get metadata, astmetad is not running")
+        except JSONDecodeError:
+            LOGGER.error("Could not decode JSON metadata.")
+        if self._metadata_message is not None:
+            self._astmetad_received.set()
```
Comment (trickeydan) [inline sr/robot3/astoria.py]: It's not possible for `message` to be `None` here. Pydantic will either return a parsed and valid `MetadataManagerMessage` OR throw a `pydantic.ValidationError`.

```suggestion
        try:
            message = parse_obj_as(MetadataManagerMessage, loads(msg.payload))
            if message.status == MetadataManagerMessage.Status.RUNNING:
                LOGGER.debug("Received metadata")
                self._metadata_message = message
                self._astmetad_received.set()
            else:
                LOGGER.warn("Cannot get metadata, astmetad is not running")
        except JSONDecodeError:
            LOGGER.error("Could not decode JSON metadata.")
        except pydantic.ValidationError as e:
            LOGGER.error(f"Unable to parse metadata message: {e}")
            
```

File: sr/robot3/astoria.py [hunk 5]
```diff
@@ -28,126 +31,197 @@ class GetMetadataResult(NamedTuple):
...
+            message = parse_obj_as(ProcessManagerMessage, loads(msg.payload))
+            if message.status == ProcessManagerMessage.Status.RUNNING:
+                LOGGER.debug("Received process info")
+                self._proc_message = message
+            else:
+                LOGGER.warn("Cannot get process info, astprocd is not running")
+        except JSONDecodeError:
+            LOGGER.error("Could not decode JSON metadata.")
+        if self._proc_message is not None:
+            self._astprocd_received.set()
```
Comment (trickeydan) [inline sr/robot3/astoria.py]: Same again here.

```suggestion
        try:
            message = parse_obj_as(ProcessManagerMessage, loads(msg.payload))
            if message.status == ProcessManagerMessage.Status.RUNNING:
                LOGGER.debug("Received process info")
                self._proc_message = message
                self._astprocd_received.set()
            else:
                LOGGER.warn("Cannot get process info, astprocd is not running")
        except JSONDecodeError:
            LOGGER.error("Could not decode JSON metadata.")
        except pydantic.ValidationError as e:
            LOGGER.error(f"Unable to parse process message: {e}")
            
```

File: sr/robot3/astoria.py [hunk 6]
```diff
@@ -28,126 +31,197 @@ class GetMetadataResult(NamedTuple):
...
-                path = gmc._proc_message.disk_info.mount_path
+            metadata_received = gmc.run(timeout=0.1)
+            if metadata_received:
+                if gmc._metadata_message is not None:
+                    metadata = gmc._metadata_message.metadata
+
+                if (
+                    gmc._proc_message is not None
+                    and gmc._proc_message.disk_info is not None
+                ):
```
Comment (trickeydan) [inline sr/robot3/astoria.py]:
```suggestion
                if gmc._proc_message and gmc._proc_message.disk_info:
```

File: sr/robot3/robot.py [hunk 7]
```diff
@@ -169,7 +170,9 @@ def is_arduino(cls, port: ListPortInfo) -> bool:
 
     def _init_metadata(self) -> None:
         """Fetch metadata from Astoria."""
-        self._metadata, self._code_path = GetMetadataConsumer.get_metadata()
+        (
+            self._metadata, self._code_path,
+        ) = GetMetadataConsumer.get_metadata(self._mqtt)
```
Comment (trickeydan) [inline sr/robot3/robot.py]:
```suggestion
        self._metadata, self._code_path = GetMetadataConsumer.get_metadata(
            self._mqtt,
        )
```

A linter isn't there to get in your way, it's there to help to remind you to make code readable. If you get up in a situation where you can't get the linter to pass in a readable way, it's better to add a per-line ignore.

File: sr/robot3/robot.py [hunk 8]
```diff
@@ -370,39 +373,36 @@ def wait_start(self) -> None:
...
-            ),
-        )
+        self.power_board.piezo.buzz(timedelta(seconds=0.1), Note.A6)
+        counter = 0
+        led_state = False
+        _ = self.power_board.start_button.is_pressed
+        while (
+            not self.power_board.start_button.is_pressed
+            and not start_event.is_set()
+        ):
```
Comment (trickeydan) [inline sr/robot3/robot.py]:
```suggestion
        while not self.power_board.start_button.is_pressed and \
            not start_event.is_set():
```

A continuation line is much better here.

File: sr/robot3/mqtt.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+            return
+
+        if auto_prefix_topic and self.topic_prefix:
+            topic = f"{self.topic_prefix}/{topic}"
+        try:
+            self._client.publish(topic, payload=payload, retain=retain, qos=1)
+        except ValueError:
+            raise ValueError(f"Cannot publish to MQTT topic: {topic}")
+
+    def __on_connect(
```
Comment (trickeydan) [inline sr/robot3/mqtt.py]:
```suggestion
    def _on_connect(
```

It's unusual to need `__` as "private" functions. Use it sparingly and only where absolutely necessary :)

File: sr/robot3/mqtt.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+        self._subscribe(full_topic, callback)
+
+    def _subscribe(
+        self,
+        topic: str,
+        callback: Callable[[mqtt.Client, Any, mqtt.MQTTMessage], None],
+    ) -> None:
+        LOGGER.debug(f"Subscribing to {topic}")
+        self._client.message_callback_add(topic, callback)
+        self._client.subscribe(topic, 1)
```
Comment (trickeydan) [inline sr/robot3/mqtt.py]:
```suggestion
        self._client.subscribe(topic, qos=1)
```

**Review by trickeydan — CHANGES_REQUESTED** [review]
Nice one! Thanks for working on this. 🎉 

I've tested this against a local version of Astoria and it works. Some issues with the implementation, but nothing major. See inline comments :)

I think we will aim to develop something more permanent for next year so that we don't have to maintain two "clients" to Astoria, but I think this will iron out the remaining issues in the kit to merely be technical debt rather than a user-facing issue.

File: sr/robot3/astoria.py (hunk 3 above)
Comment (WillB97) [reply sr/robot3/astoria.py:45]: The callbacks each only access their own attributes so there is no multiple access between callbacks and callbacks are unregistered before the main thread accesses the attributes.

Comment (WillB97) [reply sr/robot3/astoria.py:45]: `loop_start` just runs the blocking function in a background thread:

https://github.com/eclipse/paho.mqtt.python/blob/master/src/paho/mqtt/client.py#L1787
https://github.com/eclipse/paho.mqtt.python/blob/master/src/paho/mqtt/client.py#L3590

**Review by trickeydan — APPROVED** [review]

**Comment by WillB97** [comment]
This actually merged @ e2cef81, github seems to have glitched.

**Comment by trickeydan** [comment]
Screenshot 2022-11-19 at 20 22 31

much glitch lol

**Pull request closed without merging**